### PR TITLE
Update mechanism to check installed cpp extensions

### PR DESCRIPTION
**Pull request by xadupre**
**Description:**
ORTModule requires cpp extensions to be installed. onnxruntime assumes they are compiled in a subfolder and import them from this location after they were compiled. This PR introduces changes to check if these extensions are installed as a reguler package in case they are not found in the expected subfolder.

**Motivation and Context**
The cpp extensions cannot be easily installed on Windows. due to long path name, a compilation error may happen. In that case, the fallback meachnism is to manually compile every cpp extension and to install them.

### CODE REVIEW

**Review by thiagocrepaldi — COMMENTED** [review]
Should this PR also make changes to the existing packaging mechanism to pre-pack such extensions and/or update documentation insttruction how to pre-install them?

**Comment by xadupre** [comment]
The current change is compatible with the current installation and a manually installation as a regular python package. It would be more simple to have these extensions on pypi but they depend on onnxruntime and pytorch versions. It should be better just to avoid compiling the extensions. But turning them into a package can be the topic of another PR.

**Comment by thiagocrepaldi** [comment]
> The current change is compatible with the current installation and a manually installation as a regular python package. It would be more simple to have these extensions on pypi but they depend on onnxruntime and pytorch versions. It should be better just to avoid compiling the extensions. But turning them into a package can be the topic of another PR.

https://github.com/microsoft/onnxruntime/pull/8027 introduced PyTorch C++ extensions as pre-compiled binaries originally, as you envision. However our architect had concerns about ABI compatibility, which were never addressed. The concerns states that even though user can have the same pytorch version (say 1.9.1), it doesn't mean it will be necessarily compatible with a PyTorch 1.9.1 compiled during ONNX Runtime packaging.

We may use different compilers, compiler versions, dependencies versions during linking or maybe different build flags for pytorch source code. With all these variants, the same pytorch version could lead to incompatible binaries with the precompiled extensions.

At the time, even though our CI tests never caught such corner cases, our architects preferred to add this as a manual step until we find a way to validate ABI compatibility in runtime. When ABI is compatible, we could use the prepacked binaries, saving time. Otherwise, we would recompile from scratch using user's environment.

Adding a shortcut for precompiled packages without addressing the ABI concerns that drove development towards the post install approach sounds risky.

If long paths are the reason for the PR, have you considered using a different/custom path instead? ORTModule could try to load its torch c++ extensions from more than one place, but still compiling it as a post install step to prevent the ABI issues.

**Comment by thiagocrepaldi** [comment]
Please add *component:training-frontend* as *label* for frontend PRs

File: orttraining/orttraining/python/training/optim/fused_adam.py [hunk 1]
```diff
@@ -80,7 +80,10 @@ def __init__(self,
         # Skip buffer
         self._dummy_overflow_buf = torch.cuda.IntTensor([0])
 
-        from onnxruntime.training.ortmodule.torch_cpp_extensions import fused_ops
+        try:
+            from onnxruntime.training.ortmodule.torch_cpp_extensions import fused_ops
+        except ImportError:
+            import fused_ops
```
Comment (yuslepukhin) [inline orttraining/orttraining/python/training/optim/fused_adam.py:86]: Can this be a separate function?

Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/optim/fused_adam.py:86]: Good point. I was drafting a poc to refactor like this using https://docs.python.org/3/library/importlib.html#importlib.import_module

Something like
```
def import_from_onnxruntime_or_environment(name):
    ortmodule_path='onnxruntime.training.ortmodule.torch_cpp_extensions'
    try:
        import importlib
        mylib = importlib.import_module('.', ortmodule_path+'.'+name)
    except ModuleNotFoundError:
        print(name)
        mylib = importlib.import_module('.', name)
```

But although it imported the module, it wasn't defined. Some more coding must be needed to achieve this nicer refactoring

Comment (xadupre) [reply orttraining/orttraining/python/training/optim/fused_adam.py:86]: Why not something like:

```
def get_fused_ops_extension():
        try:
            from onnxruntime.training.ortmodule.torch_cpp_extensions import fused_ops
        except ImportError:
            # print, warnings, logging?
            import fused_ops
        return fused_ops

fused_ops = get_fused_ops_extension()
```

I had issues compiling these extensions on Windows (long path + unable to import them when copied at the expected location), and on Linux CPU (it complains about CUDA home if it is compiled with install.py).
